test/integration/apiserver/apply/apply_test.go

```diff
@@ -1043,3 +1048,307 @@ func TestClearManagedFieldsWithUpdate(t *testing.T) {
+        exitCode: 255
+        finishedAt: "2019-07-08T09:39:09Z"
+        reason: Error
+        startedAt: "2019-07-08T09:38:54Z"
+    name: name
+    ready: true
+    restartCount: 6
+    state:
+      running:
+        startedAt: "2019-07-08T09:41:59Z"
+  hostIP: 10.0.0.1
+  phase: Running
+  podIP: 10.0.0.1
+  qosClass: BestEffort
+  startTime: "2019-07-08T09:31:18Z"
+`)
+
+func decodePod(podBytes []byte) *v1.Pod {
+	pod := v1.Pod{}
+	err := yaml.Unmarshal(podBytes, &pod)
+	if err != nil {
+		panic(err)
+	}
+	return &pod
+}
+
+func encodePod(pod *v1.Pod) []byte {
+	podBytes, err := yaml.Marshal(pod)
+	if err != nil {
+		panic(err)
+	}
+	return podBytes
+}
+
+func BenchmarkNoServerSideApply(b *testing.B) {
+	defer featuregatetesting.SetFeatureGateDuringTest(b, utilfeature.DefaultFeatureGate, genericfeatures.ServerSideApply, false)()
+
+	_, client, closeFn := setup(b)
+	defer closeFn()
+	flag.Lookup("v").Value.Set("0")
+
+	pod := decodePod(podBytes)
+
+	// Create pod for repeated-updates
+	pod.Name = "repeated-pod"
+	_, err := client.CoreV1().RESTClient().Post().
+		Namespace("default").
+		Resource("pods").
+		SetHeader("Content-Type", "application/yaml").
+		Body(encodePod(pod)).Do().Get()
+	if err != nil {
+		b.Fatalf("Failed to create object: %v", err)
+	}
+
+	b.Run("Get1", benchGetPod(client, *pod, 1))
+	b.Run("Get20", benchGetPod(client, *pod, 20))
+	b.Run("Get200", benchGetPod(client, *pod, 200))
+	b.Run("Get2000", benchGetPod(client, *pod, 2000))
+
+	b.Run("RepeatedUpdates", benchRepeatedUpdate(client, "repeated-pod"))
+	b.Run("Post1", benchPostPod(client, *pod, 1))
+	b.Run("Post10", benchPostPod(client, *pod, 10))
+	b.Run("Post50", benchPostPod(client, *pod, 50))
+}
+
+func BenchmarkServerSideApply(b *testing.B) {
+	defer featuregatetesting.SetFeatureGateDuringTest(b, utilfeature.DefaultFeatureGate, genericfeatures.ServerSideApply, true)()
+
+	_, client, closeFn := setup(b)
+	defer closeFn()
+	flag.Lookup("v").Value.Set("0")
+
+	pod := decodePod(podBytes)
+
+	// Create pod for repeated-updates
+	pod.Name = "repeated-pod"
+	_, err := client.CoreV1().RESTClient().Post().
+		Namespace("default").
+		Resource("pods").
+		SetHeader("Content-Type", "application/yaml").
+		Body(encodePod(pod)).Do().Get()
+	if err != nil {
+		b.Fatalf("Failed to create object: %v", err)
+	}
+
+	b.Run("Get1", benchGetPod(client, *pod, 1))
+	b.Run("Get20", benchGetPod(client, *pod, 20))
+	b.Run("Get200", benchGetPod(client, *pod, 200))
+	b.Run("Get2000", benchGetPod(client, *pod, 2000))
+
+	b.Run("RepeatedUpdates", benchRepeatedUpdate(client, "repeated-pod"))
+	b.Run("Post1", benchPostPod(client, *pod, 1))
+	b.Run("Post10", benchPostPod(client, *pod, 10))
+	b.Run("Post50", benchPostPod(client, *pod, 50))
+}
+
+func benchPostPod(client kubernetes.Interface, pod v1.Pod, parallel int) func(*testing.B) {
+	return func(b *testing.B) {
+		b.ResetTimer()
+		b.ReportAllocs()
+		for i := 0; i < b.N; i++ {
+			c := make(chan error)
+			for j := 0; j < parallel; j++ {
```

Comment:
Why do things in parallel?

Reply:
We were suspecting this was making things even worse (closer to real-life system)

Reply:
I'm not sure if it's necessary to capture that in the benchmark, but I guess it doesn't hurt...